rfc-171-remove-legacy-browser-js-support.md

```diff
@@ -0,0 +1,128 @@
+
+At the time of writing the service manual (which was last updated in 2022) says "You will still need to support Internet Explorer 11 if your analytics data shows at least 2% of your users arriving at the service's start page are using it." It [recommends testing](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices) only in Edge of all the Microsoft browsers.
+
+The Design System team have updated `govuk-frontend` to use [ES modules in version 5](https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/#javascript), released in 2023. [ES modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) are not recognised by legacy browsers, notably Internet Explorer 11 and below. These browsers are no longer supported by Microsoft. We need to keep up to date with `govuk-frontend`, which means deciding how we approach JS support for IE11. The Design System has their own definitions for [browser support](https://frontend.design-system.service.gov.uk/browser-support/).
+
+The presence of any modern JS such as ES modules will cause errors for older browsers. This is because they will attempt to parse all JS before running it, encounter unfamiliar syntax, and error. No JS will run on older browsers once any modern JS is introduced, which means that any download of JS will be wasted.
+
+## Proposal
+
+The proposal is to embrace modern JS so that legacy browsers do not download any JS but see the non-JS version of GOV.UK. This should still provide a usable experience of the site as the site is built using the principle of progressive enhancement. This will also reduce page weight for those browsers and allow us to embrace modern JS techniques, leading to further code improvements in future.
+
+We should implement ES modules initially, with further work done later to potentially use other modern JS techniques. This will be done on a problem solving basis rather than upgrading everything - we will identify issues and look at how modern JS could solve them.
+
+Legacy browsers will not download any JS imported using the modules syntax. This would work as follows:
+
+- the HTML includes a JS file that imports code from other JS files using module includes
+- most browsers download and execute this code as normal, pulling in the code from the other JS files
+- legacy browsers download this one file but don't recognise the imports and therefore don't import anything else, resulting in a minimal JS download weight and a non-JS version of GOV.UK
+
+In practice this looks something like this.
+
+```HTML
+<script type="module" src="application.js"></script>
+```
+
+```JavaScript
+// application.js
+import './module1.js';
+import './module2.js';
+```
+
+```JavaScript
+// module1.js
+// code here will only run in modern browsers
+alert('hi');
+```
+
+Legacy browsers will only download `application.js` and no other JS files.
```

Comment:
Is the plan here not to use any bundlers? or is that undefined? Similarly with our use of modern JS syntax would we be doing ant transpiling or sticking with just what evergreen browsers support?

Do legacy browsers still load the JS if it's script type="module" ? I did think the advantage of `<script type="module">` was that it prevented legacy browsers - as seems a bit of a bummer if they get errors.

Reply:
Some legacy browsers may _download_ the `script type="module"` but [will not execute it](https://github.com/philipwalton/webpack-esnext-boilerplate/issues/1#issuecomment-445280691). There's no 'need' to use `import` within the JavaScript file to prevent them from executing.

Bear in mind that imports from within the JavaScript file will not be discoverable by the [preload scanner](https://web.dev/articles/preload-scanner) and the browser won't know they're needed until `application.js` has been downloaded and parsed.

Reply:
The aim of this approach would be to minimise the amount of JS downloaded by legacy browsers. If we just switch all of our script tags to type module, I believe they'd still download it even though they don't do anything with it. That's fine, but ultimately it would be better if they downloaded as little as possible if they're not going to use it. Will clarify.